crates/sui-indexer/src/types.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use sui_json_rpc_types::{
+    BalanceChange, ObjectChange, SuiTransaction, SuiTransactionEffects, SuiTransactionEvents,
+    SuiTransactionResponse,
+};
+use sui_types::digests::TransactionDigest;
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+
+#[derive(Debug, Clone)]
+pub struct SuiTransactionFullResponse {
+    pub digest: TransactionDigest,
+    /// Transaction input data
+    pub transaction: SuiTransaction,
+    pub effects: SuiTransactionEffects,
+    pub events: SuiTransactionEvents,
+    pub object_changes: Vec<ObjectChange>,
+    pub balance_changes: Vec<BalanceChange>,
+    pub timestamp_ms: u64,
+    pub confirmed_local_execution: bool,
+    pub checkpoint: CheckpointSequenceNumber,
+}
+
+impl TryFrom<SuiTransactionResponse> for SuiTransactionFullResponse {
+    type Error = anyhow::Error;
+
+    fn try_from(response: SuiTransactionResponse) -> Result<Self, Self::Error> {
+        let SuiTransactionResponse {
+            digest,
+            transaction,
+            effects,
+            events,
+            object_changes,
+            balance_changes,
+            timestamp_ms,
+            confirmed_local_execution,
+            checkpoint,
+            errors: _,
+        } = response;
+
+        let transaction = transaction.ok_or_else(|| {
+            anyhow::anyhow!(
```

Comment:
another argument around this is, when we see unexpected None values, this way we know which field is missing, but yeah I did not marry these 50 LOCs, and happy to be convinced. cc @patrickkuo